### PR TITLE
Refactor data insertion logic 

### PR DESCRIPTION
**Pull request by iuwqyir**
### TL;DR

Refactored data insertion process to allow the storage decide how to handle batched data. 
E.g sql databases can use transactions and concurrency whereas memory storage can do everything sequentially

### What changed?

- Introduced a new `InsertDataForBlocks` method in the `IMainStorage` interface.
- Removed individual insert methods for blocks, transactions, logs, and traces from the interface.
- Implemented `InsertDataForBlocks` in both ClickHouseConnector and MemoryConnector.
- Updated the Committer to use the new `InsertDataForBlocks` method instead of `saveDataToMainStorage`.
- Renamed insert methods in ClickHouseConnector and MemoryConnector to be private.

### CODE REVIEW

**Comment by iuwqyir** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/thirdweb-dev/insight/89?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#91** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/insight/91?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#90** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/insight/90?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#89** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/insight/89?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#88** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/insight/88?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @iuwqyir and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: internal/storage/clickhouse.go [hunk 1]
```diff
@@ -756,3 +757,77 @@ func (c *ClickHouseConnector) GetTraces(qf QueryFilter) (traces []common.Trace,
...
+	blocks := make([]common.Block, 0, len(data))
+	logs := make([]common.Log, 0)
+	transactions := make([]common.Transaction, 0)
+	traces := make([]common.Trace, 0)
+
+	for _, blockData := range data {
+		blocks = append(blocks, blockData.Block)
+		logs = append(logs, blockData.Logs...)
+		transactions = append(transactions, blockData.Transactions...)
+		traces = append(traces, blockData.Traces...)
```
Comment (AmineAfia) [inline internal/storage/clickhouse.go:772]: should we delete blockData after finishing here, as it's not needed anywhere and only blows the service memory

File: internal/storage/connector.go [hunk 2]
```diff
@@ -48,12 +48,9 @@ type IStagingStorage interface {
 }
 
 type IMainStorage interface {
-	InsertBlocks(blocks []common.Block) error
-	InsertTransactions(txs []common.Transaction) error
-	InsertLogs(logs []common.Log) error
-	InsertTraces(traces []common.Trace) error
+	InsertDataForBlocks(data []common.BlockData) error
```
Comment (AmineAfia) [inline internal/storage/connector.go:51]: nit: `InsertBlockData` is maybe a better name for this function

**Review by AmineAfia — APPROVED** [review]

File: internal/storage/connector.go (hunk 2 above)
Comment (iuwqyir) [reply internal/storage/connector.go:51]: Yeah, but it clashes with Staging interface where there is also a `InsertBlockData`. Do you have other naming suggestions for these 2?

File: internal/orchestrator/committer.go [hunk 3]
```diff
@@ -151,7 +150,7 @@ func (c *Committer) commit(blockData []common.BlockData) error {
 	log.Debug().Msgf("Committing %d blocks", len(blockNumbers))
 
 	// TODO if next parts (saving or deleting) fail, we'll have to do a rollback
-	if err := c.saveDataToMainStorage(blockData); err != nil {
+	if err := c.storage.MainStorage.InsertDataForBlocks(blockData); err != nil {
```
[the remark this answers is not in the document]
Comment (iuwqyir) [reply internal/orchestrator/committer.go:153]: Changed it

File: internal/storage/clickhouse.go (hunk 1 above)
Comment (iuwqyir) [reply internal/storage/clickhouse.go:772]: it is used for metrics. I think it's dangerous to delete data manually inside the storage layer if it doesn't have context about business logic

File: internal/storage/connector.go (hunk 2 above)
Comment (AmineAfia) [reply internal/storage/connector.go]: in staging we don't insert blocks anymore, we can change the staging function to `InsertStaginData`?

**Review by AmineAfia — APPROVED** [review]

**Comment by iuwqyir** [comment]
### Merge activity

* **Oct 10, 10:01 AM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/thirdweb-dev/insight/89).
